tasks/main.yml

```diff
@@ -17,26 +12,328 @@
+    ((__mssql_errorlog_default.stat is defined) and
+    (not __mssql_errorlog_default.stat.exists)))
+  register: __mssql_variables_evaluation
+
+- name: Set up MS SQL Server
+  command: /opt/mssql/bin/mssql-conf -n setup
+  environment:
+    - ACCEPT_EULA: "Y"
+    - MSSQL_SA_PASSWORD: "{{ mssql_password }}"
+    - MSSQL_PID: "{{ mssql_edition }}"
+  when: >
+    ("running" not in
+    ansible_facts["services"]["mssql-server.service"]["state"]) and
+    ("enabled" not in
+    ansible_facts["services"]["mssql-server.service"]["status"]) and
+    (((__mssql_errorlog_custom.stat is defined) and
+    (not __mssql_errorlog_custom.stat.exists)) or
+    ((__mssql_errorlog_default.stat is defined) and
+    (not __mssql_errorlog_default.stat.exists)))
+  register: __mssql_conf_setup
+
+- name: Configure the Microsoft SQL Server Tools repository
+  yum_repository:
+    name: packages-microsoft-com-prod
+    description: Microsoft SQL Server Tools
+    baseurl: "{{ __mssql_client_repository }}"
+    gpgcheck: true
+
+- name: Ensure that SQL Server client tools are installed
+  package:
+    name: "{{ __mssql_client_packages }}"
+    state: present
+  environment:
+    - ACCEPT_EULA: Y
+
+- name: Set a new password for the MSSQL Server sa user
+  when:
+    - __mssql_conf_setup.skipped is defined
+    - mssql_password is defined
+  block:
+    - name: Ensure that the mssql-server service is started
+      service:
+        name: mssql-server
+        state: started
+
+    - name: Check if a custom tcpport setting exist
+      command: /opt/mssql/bin/mssql-conf get network tcpport
+      ignore_errors: true
+      changed_when: false
+      register: __mssql_custom_tcp_port
+
+    - name: Check if a custom ipaddress setting exist
+      command: /opt/mssql/bin/mssql-conf get network ipaddress
+      ignore_errors: true
+      changed_when: false
+      register: __mssql_custom_ip_address
+
+    - name: >-
+        Check if the password matches the existing password.
+        When IP address and TCP port are default.
+      command: >-
+        /opt/mssql-tools/bin/sqlcmd -U sa
+        -P '{{ mssql_password }}' -Q 'SELECT @@VERSION'
+      register: __mssql_sqlcmd_query_default_settings
+      ignore_errors: true
+      changed_when: false
+      when:
+        - '"ipaddress" not in __mssql_custom_ip_address.stdout'
+        - '"tcpport" not in __mssql_custom_tcp_port.stdout'
+
+    - name: >-
+        Check the password matches the existing password.
+        When IP address is custom and TCP port is default.
+      command: >-
+        /opt/mssql-tools/bin/sqlcmd
+        -S {{ __mssql_custom_ip_address.stdout | regex_search('[^: ]*$') }}
+        -U sa -P '{{ mssql_password }}' -Q 'SELECT @@VERSION'
+      register: __mssql_sqlcmd_query_custom_ip
+      ignore_errors: true
+      changed_when: false
+      when:
+        - '"ipaddress" in __mssql_custom_ip_address.stdout'
+        - '"tcpport" not in __mssql_custom_tcp_port.stdout'
```

Comment:
I recommend doing something like this to avoid having 4 separate tasks that do the same thing:
```yaml
    - name: Check the password matches the existing password.
      vars:
        __ipaddress: "{{ __mssql_custom_ip_address.stdout | regex_search('[^: ]*$') }}"
        __tcpport: "{{ __mssql_custom_tcp_port.stdout | regex_search('[^: ]*$') }}"
        __s_arg: "{{ (__ipaddress or __tcpport) | ternary('-S', '') }}"
        __ipaddress_arg: "{{ __ipaddress if __ipaddress else 'localhost' if __tcpport else '' }}"
    command: >-
      /opt/mssql-tools/bin/sqlcmd
      {{ __s_arg }}
      {{ __ipaddress_arg }}{{ ',' if __tcpport else '' }}{{ __tcpport if __tcpport else '' }}
      -U sa -P {{ mssql_password | quote }} -Q 'SELECT @@VERSION'
```

I find it useful to write test playbooks to try stuff out:
```yaml
- hosts: localhost
  gather_facts: false
  vars:
    mypassword: "this is a <' bad '> password!"
  tasks:
    - name: debug
      vars:
        __ipaddress: "{{ item.ip_address_stdout | regex_search('[^: ]*$') }}"
        __tcpport: "{{ item.tcp_port_stdout | regex_search('[^: ]*$') }}"
        __s_arg: "{{ (__ipaddress or __tcpport) | ternary('-S', '') }}"
        __ipaddress_arg: "{{ __ipaddress if __ipaddress else 'localhost' if __tcpport else '' }}"
      command: >-
        echo /opt/mssql-tools/bin/sqlcmd
        {{ __s_arg }}
        {{ __ipaddress_arg }}{{ ',' if __tcpport else '' }}{{ __tcpport if __tcpport else '' }}
        -U sa -P {{ mypassword | quote }} -Q 'SELECT @@VERSION'
      loop:
        - ip_address_stdout: ""
          tcp_port_stdout: ""
        - ip_address_stdout: "ipaddress : 127.0.0.1"
          tcp_port_stdout: ""
        - ip_address_stdout: ""
          tcp_port_stdout: "tcpport : 4999"
        - ip_address_stdout: "ipaddress : 127.0.0.1"
          tcp_port_stdout: "tcpport : 4999"
```
run like this
```
ANSIBLE_STDOUT_CALLBACK=debug ansible-playbook -vv mssql-test.yml
```
example output:
```
changed: [localhost] => (item={'ip_address_stdout': 'ipaddress : 127.0.0.1', 'tcp_port_stdout': 'tcpport : 4999'}) => {
    "ansible_loop_var": "item",
    "changed": true,
    "cmd": [
        "echo",
        "/opt/mssql-tools/bin/sqlcmd",
        "-S",
        "127.0.0.1,4999",
        "-U",
        "sa",
        "-P",
        "this is a <' bad '> password!",
        "-Q",
        "SELECT @@VERSION"
    ],
    "delta": "0:00:00.001644",
    "end": "2021-03-05 17:29:40.176027",
    "item": {
        "ip_address_stdout": "ipaddress : 127.0.0.1",
        "tcp_port_stdout": "tcpport : 4999"
    },
    "rc": 0,
    "start": "2021-03-05 17:29:40.174383"
}

STDOUT:

/opt/mssql-tools/bin/sqlcmd -S 127.0.0.1,4999 -U sa -P this is a <' bad '> password! -Q SELECT @@VERSION
```
also note the use of `quote` for the password to ensure is is quoted correctly.

Reply:
I adore your solution and I've implemented it with some changes: I need to change the regex to either match the IP address or match nothing.